controllers/pod_controller.go

```diff
@@ -0,0 +1,190 @@
+type PodReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile reconciles Pods with the `instanceLabel` set.
+// It tracks state changes in an annotation on the pod and calculates the timing metrics on pod completion or timeout.
+// Completed pods, pods with no timeout annotation, and timed out pods are deleted.
+func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx)
+
+	pod := &corev1.Pod{}
+	err := r.Client.Get(ctx, req.NamespacedName, pod)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	if !pod.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Object is in the process of being deleted.
+		return ctrl.Result{}, nil
+	}
+
+	instance := pod.GetLabels()[instanceLabel]
+
+	timeout, err := timeoutFromPod(pod)
+	if err != nil {
+		l.Error(err, "Pod has no valid timeout annotation, deleting immediately")
+	}
+
+	timedOut := podHasReachedTimeout(*pod, timeout)
+	state := podstate.State(*pod)
+	l.Info("Pod is in state", "state", state, "timedOut", timedOut)
+	if timedOut || state == podstate.PodCompleted {
+		if err := recordSchedulingMetrics(l, instance, pod, timedOut); err != nil {
+			l.Error(err, "Failed to record metrics")
+		}
+		return ctrl.Result{}, r.Client.Delete(ctx, pod)
+	}
+
+	if err := trackState(pod, state); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	deadline := pod.CreationTimestamp.Add(timeout)
+	return ctrl.Result{Requeue: true, RequeueAfter: time.Until(deadline)}, r.strategicMergePatch(ctx, pod, stateTrackingPatchFromPod(*pod))
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PodReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	filter, err := predicate.LabelSelectorPredicate(v1.LabelSelector{
+		MatchExpressions: []v1.LabelSelectorRequirement{
+			{
+				Key:      instanceLabel,
+				Operator: v1.LabelSelectorOpExists,
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Pod{}, builder.WithPredicates(filter)).
+		Complete(r)
+}
+
+func (r *PodReconciler) strategicMergePatch(ctx context.Context, obj client.Object, patch map[string]any) error {
+	jp, err := json.Marshal(patch)
+	if err != nil {
+		return fmt.Errorf("failed to marshal patch: %w", err)
+	}
+	return r.Client.Patch(ctx, obj, client.RawPatch(types.StrategicMergePatchType, jp))
+}
+
+func recordSchedulingMetrics(l logr.Logger, instance string, pod *corev1.Pod, timedOut bool) error {
+	tr, err := getTrackedStates(pod)
+	if err != nil {
+		return fmt.Errorf("failed to get state tracking timestamps: %w", err)
+	}
+
+	createdTime, ok := tr[podstate.PodCreated]
+	if !ok {
+		return fmt.Errorf("pod created time not found")
+	}
+
+	metricLabels := prometheus.Labels{"namespace": pod.Namespace, "name": instance}
+
+	cm := podTimeCompleted.MustCurryWith(metricLabels)
```

Comment:
Are you 100% sure this will never fail? We should be very careful with `MustX` functions..

Reply:
Yes I'm sure. It does not change dynamically. Only with the metric definition. Which if changed would crash the test suite.